packages/consensus/src/types/utreexo.cairo

```diff
@@ -72,6 +75,66 @@ pub trait UtreexoAccumulator {
     fn delete_batch(ref self: UtreexoState, proof: @UtreexoBatchProof);
 }
 
+fn parent_poseidon(left: felt252, right: felt252) -> felt252 {
+    let mut arr: Array<felt252> = array![];
+    arr.append(left);
+    arr.append(right);
+    let span: Span<felt252> = arr.span();
+
+    let mut hash = PoseidonTrait::new().update(left).update(right);
+    println!("hash: {:?}", hash);
+    let hash_felt252 = hash.update(poseidon_hash_span(span)).finalize();
+    println!("hash felt: {}", hash_felt252);
+
+    hash_felt252
+}
+
+pub impl UtreexoAccumulatorImpl of UtreexoAccumulator {
+    //  n ← leaf                . n is initially the leaf to add
+    //  h ← 0                   . height is initially 0
+    //  r ← acc[h]              . r is the lowest root
+    //  while r 6= ∅ do         . loop until an empty root
+    //     n ← Parent(r, n)
+    //     acc[h] ← ∅
+    //     h ← h + 1
+    //     r ← acc[h]
+    //   acc[h] ← n
+
+    fn add(ref self: UtreexoState, outpoint: OutPoint) {
+        let mut new_leaf: OutPoint = outpoint;
+
+        let mut new_leaf_hash: felt252 = PoseidonTrait::new().update_with(new_leaf).finalize();
+        println!("new leaf hash: {}", new_leaf_hash);
+        let mut height = 0;
+
+        while let Option::Some(root) = *self.roots[height] {
+            // self.roots[height] = Option::None; ??
+            println!("root[{height}]: {}", root);
+            new_leaf_hash = parent_poseidon(root, new_leaf_hash);
```

Comment:
You also have to set the current root to empty.
Since we cannot modify array elements, you'd need to iterate over all roots and produce a new array of roots